### PR TITLE
Add datadog plugin

### PR DESCRIPTION
**Pull request by applitect**
What can I say? It works, has been tested. Lots of unit tests. Will get documentation added in next pull request.

### CODE REVIEW

File: server/src/main/java/forklift/ForkliftPropFile.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+     * code in forklift can then access the properties via System.getProperty
+     *
+     * @param forkliftConfigFile - The filename of the properties file on the classpath or as a URL
+     */
+    public static void addPropertiesFromFile(String forkliftConfigFile) {
+        Properties p = new Properties();
+        URL url = findConfigFileURLFromSystemProperties(ForkliftServer.class.getClassLoader(), forkliftConfigFile);
+        if (url != null) {
+            try (BufferedReader in = new BufferedReader(new InputStreamReader(
+                    url.openStream()))) {
```
Comment (Kuroshii) [inline server/src/main/java/forklift/ForkliftPropFile.java:64]: nit: awkward newline

File: server/src/test/resources/example.properties [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# An example prooperties file for some forklift props
+
+# Properties for datadog keys
+datadog.api.key:e4335ba2030485b6f650ceac0bbd9531
```
Comment (Kuroshii) [inline server/src/test/resources/example.properties:4]: I assume this is just random data?

File: plugins/datadog/src/main/java/forklift/datadog/DatadogCollector.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+ * A plug-in that allows tracking portions of the forklift lifecycle directly in Datadog.
+ *
+ * @author David Thompson
+ *
+ * Create datadog metrics named trace.forklift.count trace.forklift.timer
+ *
+ * Include tags: consumer-name, lifecycle, environment, host
+ *
+ */
+public class DatadogCollector extends SimpleCollector {
```
[the remark this answers is not in the document]
Comment (applitect) [reply plugins/datadog/src/main/java/forklift/datadog/DatadogCollector.java:28]: Agreed, but I'll leave this for someone else to refactor later if they want. Just trying to get this in to the version for our specific use case now.

File: plugins/datadog/src/main/java/forklift/datadog/SimpleCollector.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        return timerStart(consumerName, lifecycle);
+    }
+
+    Timer.Sample timerStart(String consumerName, String lifecycle) {
+        if (consumerName == null || consumerName.length() == 0)
+            return null;
+        if (lifecycle == null || lifecycle.length() == 0)
+            return null;
+
+        Map<String, Timer.Sample> queueTimers = timers.get(consumerName);
```
[the remark this answers is not in the document]
Comment (applitect) [reply plugins/datadog/src/main/java/forklift/datadog/SimpleCollector.java]: Rewritten and thanks for pointing this out. I hadn't used this method before.

File: plugins/datadog/src/main/java/forklift/datadog/SimpleCollector.java [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        if (queueTimers == null) {
+            queueTimers = new HashMap<>();
+            timers.put(consumerName, queueTimers);
+        }
+        // In order to make this thread-safe, we need to add the thread id.
+        // Since the lifecycle in forklift is guaranteed to run on a single thread
+        // then we can guarantee that start and stop will live on the same
+        // thread, thus the same thread id.
+        String threadLifecycle = lifecycle + Thread.currentThread().getId();
+        Timer.Sample sampler = queueTimers.get(threadLifecycle);
```
[the remark this answers is not in the document]
Comment (applitect) [reply plugins/datadog/src/main/java/forklift/datadog/SimpleCollector.java]: Agreed this was overly complicated. Now cleaned up to use a composite key.

File: plugins/datadog/src/main/java/forklift/datadog/SimpleCollector.java [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+        return timer.count();
+    }
+
+    /**
+     * Looking at system properties that have been set, determine if the lifecycle
+     * metering should be turned off.
+     * @param consumerName - The name of the queue, topic, stream, etc.
+     * @param lifecycle - The lifecycle step in forklift
+     * @param propValue - The property value if set can turn off a timer
```
[the remark this answers is not in the document]
Comment (applitect) [reply plugins/datadog/src/main/java/forklift/datadog/SimpleCollector.java]: Renamed to be propertyPrefix.

File: server/src/test/resources/example.properties (hunk 2 above)
Comment (applitect) [reply server/src/test/resources/example.properties:4]: Yep, just an example. Not a real key.

File: plugins/datadog/build.gradle [hunk 7]
```diff
@@ -0,0 +1,7 @@
+dependencies {
+    api project(':forklift-core'),
+    'io.micrometer:micrometer-registry-prometheus:latest.release',
```
[the remark this answers is not in the document]
Comment (applitect) [reply plugins/datadog/build.gradle]: Probably a cut and paste. Isn't even needed since micrometer-datadog will add micrometer-core.

**Pull request merged**